graph/patterns/alternate-key.md

```diff
@@ -0,0 +1,48 @@
+# Alternate Key Pattern
+
+Microsoft Graph API Design Pattern
+
+*The Alternate Key Pattern provides the ability to query for a single, specific entity identifiable through an alternative attribute (called key) that is not its unique identifier*
+
+## Problem
+--------
+
+All entities in our system are identified by an UUID - which guarantees uniqueness. Often though, that same entity can also be uniquely identified by an alternative, more convenient attribute.
+
+Take a look at the `user` entity: while the UUID remains a perfectly valid way to get the entity details, the `email` address is also an unique attribute that could be used to identify it.
+
+While it is still possible to use the oData filter, such as
+
+`serviceRoot/Users?$filter=email eq 'hello@microsoft.com'`, the returned result is wrapped in an array that needs to be unpacked.
+
+
+## Solution
+--------
+
+oData offers entity addressing via an alternate key using the same parentheses-style convention as for the canonical key, with one difference: single-part alternate keys MUST specify the key property name to unambiguously determine the alternate key.
+
+http://host/service/Employees(0) - Retrieves the employee with ID = 0
+http://host/service/Employees(email='hello@microsoft.com') Retrieves the employee with the email matching `hello@microsoft.com`
+
+## When to Use this Pattern
+------------------------
+
+This pattern works when the alternate key is good enough to identify a single entity that is properly namespaced; while it does not work if the resultset has more than one element.
```

Comment:
What do you mean by "that is properly namespaced"? I would remove it

I wonder whether we should also say that this is a good pattern to adopt when customers/developers ask for alternative keys to make their lives easier.

Reply:
I have probably chosen an unhappy wording for this.

What I was trying to communicate is that you need to be in a context that is narrow enough where the alternate key returns a single entity - I do see how saying "namespaced" makes the things confusing. I have reworded in 8f4966c, let me know if that is better.